drafts/arcalinea-tx-expiry/draft.rst

```diff
@@ -0,0 +1,76 @@
+
+Advantages include improving performance by removing transactions unlikely to be mined, and potentially simplifying bidirectional payment channels by reducing the need to store and compress revocations for past states, since transactions not committed to the chain could expire and become invalid after a period of time.
+
+Specification
+===============
+
+Transactions will have a new field, nExpiryTime, which will set the block height or UNIX timestamp after which transactions will be removed from the mempool if they have not been mined.
+
+The data type for nExpiryTime will be uint32_t, conforming to the structure of nLockTime. If used in combination with nLockTime, both nLockTime and nExpiryTime must be the same unit (either block height or timestamp).
+
+For the example below, the last block that the transaction below could possibly be included in is 3539. After that, it will be removed from the mempool.
+
+````
+"txid": "17561b98cc77cd5a984bb959203e073b5f33cf14cbce90eb32b95ae2c796723f",
+"version": 3,
+"locktime": 2089,
+"blockexpiry": 3539,
+```
+
+Default: nExpiryTime defaults to 0, which means transaction does not expire, preserving current behavior.
+Minimum: When nExpiryTime is set, it must be at least 60 blocks greater than the current blockheight, or 2 hrs greater than LOCKTIME_MEDIAN_TIME_PAST. This is because of the safety thresholds specified below.
+Maximum: No maximum specified, although based on the highest timestamp available for uint32 numbers, nExpiryTime becomes meaningless after the year 2106.
+
+Timeline of expiration safety thresholds
+-----------------------------------------
+
+2 hours before a transaction is set to expire, other nodes stop relaying it.
+1 hour before expiring, CreateNewBlock does not include the tx anymore. This is in case mining software changes the timestamp afterwards, resulting in the mining of an invalid block that contains expired transactions.
+1 hour after expiring, wallet marks the transaction as expired and frees the coins.
+
+Another reason to set a minimum expiration period is to mitigate the use of this feature as a spam mechanism.
+
+Upon block reorganization
+--------------------------
+
+Transactions that are confirmed close to the end of their expiry period may be dropped from the mempool upon a block reorg, which could leave dependent transactions stranded in the mempool. Therefore, each time an expired transaction is removed from the mempool, a check must be added to remove its dependent transactions as well.
+
+Wallet behavior and UI
+-----------------------
+
+With the addition of this feature, zero-confirmation transactions with an expiration block height set will have even less guarantee of inclusion. This means that UIs and services must never rely on zero-confirmation transactions in Zcash.
+
+Wallet should notify user of expired transactions that must be re-sent.
```

Comment:
In addition to this, can expired transactions be prevented? Error prevention, not resolution, is the best way to handle errors. 

If a wallet predict which transactions are too large or insufficiently incentivized, I suggest they do that in a conservative manner. I know that this depends on the state of the network and require some technical work for this to happen in a meaningful way without many false positives.